ocaml/testsuite/tests/typing-layouts-arrays/test_gen_u_array.ml

```diff
@@ -0,0 +1,958 @@
+  let fill a ofs len v = fill a ofs len (of_boxed v)
+  let iter f t = iter (fun v -> f (to_boxed v)) t
+  let iteri f t = iteri (fun i v -> f i (to_boxed v)) t
+  let map f t = map (fun v -> of_boxed (f (to_boxed v))) t
+  let map_inplace f t = map_inplace (fun v -> of_boxed (f (to_boxed v))) t
+  let mapi f t = mapi (fun i v -> of_boxed (f i (to_boxed v))) t
+  let mapi_inplace f t = mapi_inplace (fun i v -> of_boxed (f i (to_boxed v))) t
+  let fold_left f acc t = fold_left (fun acc v -> f acc (to_boxed v)) acc t
+  let fold_right f t acc = fold_right (fun v acc -> f (to_boxed v) acc) t acc
+
+  let iter2 f a b = iter2 (fun v1 v2 -> f (to_boxed v1) (to_boxed v2)) a b
+  let map2 f a b = map2 (fun v1 v2 -> of_boxed (f (to_boxed v1) (to_boxed v2))) a b
+  let for_all f t = for_all (fun v -> f (to_boxed v)) t
+  let exists f t = exists (fun v -> f (to_boxed v)) t
+  let mem v t = mem (of_boxed v) t
+
+  let find_index f t = find_index (fun v -> f (to_boxed v)) t
+  let find_opt f t =
+    match find_index f t with
+    | None -> None
+    | Some idx -> Some (get t idx)
+  let find_map f t = find_map (fun v -> f (to_boxed v)) t
+  let find_mapi f t = find_mapi (fun i v -> f i (to_boxed v)) t
+
+  let sort f t = sort (fun a b -> f (to_boxed a) (to_boxed b)) t
+  let stable_sort f t = stable_sort (fun a b -> f (to_boxed a) (to_boxed b)) t
+  let fast_sort f t = fast_sort (fun a b -> f (to_boxed a) (to_boxed b)) t
+
+  let map_to_array f t =
+    if length t = 0 then [||] else begin
+      let res = Array.make (length t) (f (get t 0)) in
+      iteri (fun idx v -> if idx > 0 then Array.set res idx (f v)) t;
+      res
+    end
+
+  let map_from_array f a =
+    if Array.length a = 0 then empty () else begin
+      let res = make (Array.length a) (f (Array.get a 0)) in
+      Array.iteri (fun idx v -> if idx > 0 then set res idx (f v)) a;
+      res
+    end
+
+  let unsafe_get t idx = to_boxed (unsafe_get t idx)
+  let unsafe_set t idx v = unsafe_set t idx (of_boxed v)
+  let equal = for_all2 (fun x y -> to_boxed x = to_boxed y)
+  let for_all2 f t1 t2 = for_all2 (fun a b -> f (to_boxed a) (to_boxed b)) t1 t2
+  let exists2 f t1 t2 = exists2 (fun a b -> f (to_boxed a) (to_boxed b)) t1 t2
+end
+
+module Test (A : S) : sig end = struct
+
+  (* auxiliary functions *)
+  let module I = A.I in
+  let rec check_i_upto a i =
+    if i >= 0 then begin
+      assert (A.get a i = I.of_int i);
+      check_i_upto a (i - 1);
+    end
+  in
+
+  let check_i a = check_i_upto a (A.length a - 1) in
+
+  let check_inval f arg =
+    match f arg with
+    | _ -> Format.printf "got here\n@?"; assert false
+    | exception (Invalid_argument _) -> ()
+    | exception _ -> assert false
+  in
+
+  (* [make] [set] [get] *)
+  let a = A.make 1000 (I.of_int 1) in
+  for i = 0 to 499 do A.set a i (I.of_int i) done;
+  let rec loop i =
+    if i >= 0 then begin
+      assert (A.get a i = (if i < 500 then I.of_int i else (I.of_int 1)));
```

Comment:
same remark: since we have access to `I.compare`, it'd be better to use it instead of the polymorphic equality